lib/auth/recordingmetadata/recordingmetadatav1/recordingmetadata.go

```diff
@@ -0,0 +1,350 @@
+type RecordingMetadataServiceConfig struct {
+	// Streamer is used to stream session events.
+	Streamer player.Streamer
+	// UploadHandler is used to upload session metadata and thumbnails.
+	UploadHandler uploadHandler
+}
+
+const (
+	// inactivityThreshold is the duration after which an inactivity event is recorded.
+	inactivityThreshold = 10 * time.Second
+
+	// maxThumbnails is the maximum number of thumbnails to store in the session metadata.
+	maxThumbnails = 1000
+)
+
+// NewRecordingMetadataService creates a new instance of RecordingMetadataService with the provided configuration.
+func NewRecordingMetadataService(cfg RecordingMetadataServiceConfig) (*RecordingMetadataService, error) {
+	if cfg.Streamer == nil {
+		return nil, trace.BadParameter("streamer is required")
+	}
+	if cfg.UploadHandler == nil {
+		return nil, trace.BadParameter("uploadHandler is required")
+	}
+
+	return &RecordingMetadataService{
+		streamer:      cfg.Streamer,
+		uploadHandler: cfg.UploadHandler,
+		logger:        slog.With(teleport.ComponentKey, "recording_metadata"),
+	}, nil
+}
+
+// ProcessSessionRecording processes the session recording associated with the provided session ID.
+// It streams session events, generates metadata, and uploads thumbnails and metadata.
+func (s *RecordingMetadataService) ProcessSessionRecording(ctx context.Context, sessionID session.ID) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	evts, errors := s.streamer.StreamSessionEvents(ctx, sessionID, 0)
+
+	var startTime time.Time
+	var lastEvent apievents.AuditEvent
+	var lastActivityTime time.Time
+
+	thumbnailInterval := 1 * time.Second
+	activeUsers := make(map[string]time.Duration)
+
+	vt := vt10x.New()
+
+	metadata := &pb.SessionRecordingMetadata{
+		Events: make([]*pb.SessionRecordingEvent, 0),
+	}
+
+	addInactivityEvent := func(start, end time.Time) {
+		if end.IsZero() {
+			return
+		}
+
+		inactivityStart := durationpb.New(start.Sub(startTime))
+		inactivityEnd := durationpb.New(end.Sub(startTime))
+
+		metadata.Events = append(metadata.Events, &pb.SessionRecordingEvent{
+			StartOffset: inactivityStart,
+			EndOffset:   inactivityEnd,
+			Event: &pb.SessionRecordingEvent_Inactivity{
+				Inactivity: &pb.SessionRecordingInactivityEvent{},
+			},
+		})
+	}
+
+	sampler := newThumbnailBucketSampler(maxThumbnails, thumbnailInterval)
+
+	recordThumbnail := func(start time.Time) {
+		state := vt.DumpState()
+
+		sampler.add(&state, start)
+	}
+
+loop:
+	for {
+		select {
+		case evt, ok := <-evts:
+			if !ok {
+				break loop
+			}
+
+			lastEvent = evt
+
+			switch e := evt.(type) {
+			case *apievents.DatabaseSessionStart, *apievents.WindowsDesktopSessionStart:
+				// Unsupported session recording types
+				return nil
+
+			case *apievents.Resize:
+				size, err := session.UnmarshalTerminalParams(e.TerminalSize)
+				if err != nil {
+					return trace.Wrap(err, "failed to parse terminal size %q for session %v", e.TerminalSize, sessionID)
```

Comment:
Nit: when wrapping errors in Go, I tend to prefer avoiding words like "failed" or "error" as when it's been wrapped a few times, you end up with a message that repeats those words.

```suggestion
					return trace.Wrap(err, "parsing terminal size %q for session %v", e.TerminalSize, sessionID)
```